// @ika.xyz/sdk/dist/cjs/generated/ika_common/advance_epoch_approver.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var advance_epoch_approver_exports = {};
__export(advance_epoch_approver_exports, {
  AdvanceEpochApprover: () => AdvanceEpochApprover,
  approveAdvanceEpochByWitness: () => approveAdvanceEpochByWitness,
  assertAllWitnessesApproved: () => assertAllWitnessesApproved,
  create: () => create,
  destroy: () => destroy,
  newEpoch: () => newEpoch
});
module.exports = __toCommonJS(advance_epoch_approver_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var balance = __toESM(require("./deps/sui/balance.js"));
const $moduleName = "@local-pkg/common::advance_epoch_approver";
const AdvanceEpochApprover = new import_utils.MoveStruct({ name: `${$moduleName}::AdvanceEpochApprover`, fields: {
  new_epoch: import_bcs.bcs.u64(),
  remaining_witnesses_to_approve: import_bcs.bcs.vector(import_bcs.bcs.string()),
  balance_ika: balance.Balance
} });
function create(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "u64",
    "vector<0x0000000000000000000000000000000000000000000000000000000000000001::string::String>",
    `0x0000000000000000000000000000000000000000000000000000000000000002::balance::Balance<${packageAddress}::ika::IKA>`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["newEpoch", "remainingWitnessesToApprove", "balanceIka", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "advance_epoch_approver",
    function: "create",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::advance_epoch_approver::AdvanceEpochApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "advance_epoch_approver",
    function: "new_epoch",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function assertAllWitnessesApproved(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::advance_epoch_approver::AdvanceEpochApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "advance_epoch_approver",
    function: "assert_all_witnesses_approved",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function destroy(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::advance_epoch_approver::AdvanceEpochApprover`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["self", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "advance_epoch_approver",
    function: "destroy",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function approveAdvanceEpochByWitness(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::advance_epoch_approver::AdvanceEpochApprover`,
    `${options.typeArguments[0]}`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::balance::Balance<${packageAddress}::ika::IKA>`
  ];
  const parameterNames = ["advanceEpochApprover", "_", "balanceIka"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "advance_epoch_approver",
    function: "approve_advance_epoch_by_witness",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
