// @ika.xyz/sdk/dist/cjs/generated/ika_common/bls_committee.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bls_committee_exports = {};
__export(bls_committee_exports, {
  BlsCommittee: () => BlsCommittee,
  BlsCommitteeMember: () => BlsCommitteeMember,
  CommitteeQuorumVerifiedEvent: () => CommitteeQuorumVerifiedEvent,
  contains: () => contains,
  empty: () => empty,
  isQuorumThreshold: () => isQuorumThreshold,
  isValidityThreshold: () => isValidityThreshold,
  members: () => members,
  newBlsCommittee: () => newBlsCommittee,
  newBlsCommitteeMember: () => newBlsCommitteeMember,
  quorumThreshold: () => quorumThreshold,
  totalVotingPower: () => totalVotingPower,
  validatorId: () => validatorId,
  validatorIds: () => validatorIds,
  validityThreshold: () => validityThreshold,
  verifyCertificate: () => verifyCertificate
});
module.exports = __toCommonJS(bls_committee_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var group_ops = __toESM(require("./deps/sui/group_ops.js"));
const $moduleName = "@local-pkg/common::bls_committee";
const BlsCommitteeMember = new import_utils.MoveStruct({ name: `${$moduleName}::BlsCommitteeMember`, fields: {
  validator_id: import_bcs.bcs.Address,
  protocol_pubkey: group_ops.Element
} });
const BlsCommittee = new import_utils.MoveStruct({ name: `${$moduleName}::BlsCommittee`, fields: {
  members: import_bcs.bcs.vector(BlsCommitteeMember),
  /** The aggregation of public keys for all members of the committee */
  aggregated_protocol_pubkey: group_ops.Element,
  /** Minimum signatures required for quorum (2n/3 + 1) */
  quorum_threshold: import_bcs.bcs.u64(),
  /** Minimum signatures required for validity (n/3 + 1) */
  validity_threshold: import_bcs.bcs.u64()
} });
const CommitteeQuorumVerifiedEvent = new import_utils.MoveStruct({ name: `${$moduleName}::CommitteeQuorumVerifiedEvent`, fields: {
  epoch: import_bcs.bcs.u64(),
  signer_count: import_bcs.bcs.u64()
} });
function totalVotingPower(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "total_voting_power",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function quorumThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "quorum_threshold",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function validityThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "validity_threshold",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newBlsCommitteeMember(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "0x0000000000000000000000000000000000000000000000000000000000000002::group_ops::Element<0x0000000000000000000000000000000000000000000000000000000000000002::bls12381::UncompressedG1>"
  ];
  const parameterNames = ["validatorId", "protocolPubkey"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "new_bls_committee_member",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommitteeMember`
  ];
  const parameterNames = ["member"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newBlsCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `vector<${packageAddress}::bls_committee::BlsCommitteeMember>`
  ];
  const parameterNames = ["members"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "new_bls_committee",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function empty(options = {}) {
  const packageAddress = options.package ?? "@local-pkg/common";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "empty"
  });
}
function members(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "members",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorIds(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "validator_ids",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function contains(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "validatorId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "contains",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifyCertificate(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "u64",
    "vector<u8>",
    "vector<u8>",
    "vector<u8>"
  ];
  const parameterNames = ["self", "epoch", "signature", "signersBitmap", "intentBytes"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "verify_certificate",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isQuorumThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "u64"
  ];
  const parameterNames = ["self", "signerCount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "is_quorum_threshold",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isValidityThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "u64"
  ];
  const parameterNames = ["self", "signerCount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "is_validity_threshold",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
