// @ika.xyz/sdk/dist/cjs/generated/ika_common/system_current_status_info.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var system_current_status_info_exports = {};
__export(system_current_status_info_exports, {
  SystemCurrentStatusInfo: () => SystemCurrentStatusInfo,
  create: () => create,
  currentEpoch: () => currentEpoch,
  currentEpochActiveCommittee: () => currentEpochActiveCommittee,
  isEndEpochTime: () => isEndEpochTime,
  isMidEpochTime: () => isMidEpochTime,
  nextEpochActiveCommittee: () => nextEpochActiveCommittee
});
module.exports = __toCommonJS(system_current_status_info_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var bls_committee = __toESM(require("./bls_committee.js"));
const $moduleName = "@local-pkg/common::system_current_status_info";
const SystemCurrentStatusInfo = new import_utils.MoveStruct({ name: `${$moduleName}::SystemCurrentStatusInfo`, fields: {
  current_epoch: import_bcs.bcs.u64(),
  is_mid_epoch_time: import_bcs.bcs.bool(),
  is_end_epoch_time: import_bcs.bcs.bool(),
  current_epoch_active_committee: bls_committee.BlsCommittee,
  next_epoch_active_committee: import_bcs.bcs.option(bls_committee.BlsCommittee)
} });
function create(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "u64",
    "bool",
    "bool",
    `${packageAddress}::bls_committee::BlsCommittee`,
    `0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<${packageAddress}::bls_committee::BlsCommittee>`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["currentEpoch", "isMidEpochTime", "isEndEpochTime", "currentEpochActiveCommittee", "nextEpochActiveCommittee", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "create",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function currentEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "current_epoch",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isMidEpochTime(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "is_mid_epoch_time",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isEndEpochTime(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "is_end_epoch_time",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function currentEpochActiveCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "current_epoch_active_committee",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function nextEpochActiveCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "next_epoch_active_committee",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
