// @ika.xyz/sdk/dist/cjs/generated/ika_common/upgrade_package_approver.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var upgrade_package_approver_exports = {};
__export(upgrade_package_approver_exports, {
  UpgradePackageApprover: () => UpgradePackageApprover,
  approveUpgradePackageByWitness: () => approveUpgradePackageByWitness,
  assertAllWitnessesApproved: () => assertAllWitnessesApproved,
  commit: () => commit,
  create: () => create,
  destroy: () => destroy,
  migrationEpoch: () => migrationEpoch,
  newPackageId: () => newPackageId,
  oldPackageId: () => oldPackageId
});
module.exports = __toCommonJS(upgrade_package_approver_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
const $moduleName = "@local-pkg/common::upgrade_package_approver";
const UpgradePackageApprover = new import_utils.MoveStruct({ name: `${$moduleName}::UpgradePackageApprover`, fields: {
  upgrade_cap_id: import_bcs.bcs.Address,
  remaining_witnesses_to_approve: import_bcs.bcs.vector(import_bcs.bcs.string()),
  old_package_id: import_bcs.bcs.Address,
  new_package_id: import_bcs.bcs.option(import_bcs.bcs.Address),
  migration_epoch: import_bcs.bcs.u64()
} });
function create(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "vector<0x0000000000000000000000000000000000000000000000000000000000000001::string::String>",
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u64",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["upgradeCapId", "remainingWitnessesToApprove", "oldPackageId", "migrationEpoch", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "create",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function assertAllWitnessesApproved(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "assert_all_witnesses_approved",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newPackageId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "new_package_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function oldPackageId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "old_package_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function migrationEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "migration_epoch",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function destroy(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["self", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "destroy",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function commit(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::package::UpgradeReceipt",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["self", "receipt", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "commit",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function approveUpgradePackageByWitness(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`,
    `${options.typeArguments[0]}`
  ];
  const parameterNames = ["upgradePackageApprover", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "approve_upgrade_package_by_witness",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
