// @ika.xyz/sdk/dist/cjs/generated/ika_common/validator_cap.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validator_cap_exports = {};
__export(validator_cap_exports, {
  ValidatorCap: () => ValidatorCap,
  ValidatorCommissionCap: () => ValidatorCommissionCap,
  ValidatorOperationCap: () => ValidatorOperationCap,
  VerifiedValidatorCap: () => VerifiedValidatorCap,
  VerifiedValidatorCommissionCap: () => VerifiedValidatorCommissionCap,
  VerifiedValidatorOperationCap: () => VerifiedValidatorOperationCap,
  createVerifiedValidatorCap: () => createVerifiedValidatorCap,
  createVerifiedValidatorCommissionCap: () => createVerifiedValidatorCommissionCap,
  createVerifiedValidatorOperationCap: () => createVerifiedValidatorOperationCap,
  newValidatorCap: () => newValidatorCap,
  newValidatorCommissionCap: () => newValidatorCommissionCap,
  newValidatorOperationCap: () => newValidatorOperationCap,
  validatorCommissionCapValidatorId: () => validatorCommissionCapValidatorId,
  validatorId: () => validatorId,
  validatorOperationCapValidatorId: () => validatorOperationCapValidatorId,
  verifiedValidatorCapValidatorId: () => verifiedValidatorCapValidatorId,
  verifiedValidatorCommissionCapValidatorId: () => verifiedValidatorCommissionCapValidatorId,
  verifiedValidatorOperationCapValidatorId: () => verifiedValidatorOperationCapValidatorId
});
module.exports = __toCommonJS(validator_cap_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
const $moduleName = "@local-pkg/common::validator_cap";
const ValidatorCap = new import_utils.MoveStruct({ name: `${$moduleName}::ValidatorCap`, fields: {
  id: object.UID,
  validator_id: import_bcs.bcs.Address
} });
const ValidatorOperationCap = new import_utils.MoveStruct({ name: `${$moduleName}::ValidatorOperationCap`, fields: {
  id: object.UID,
  validator_id: import_bcs.bcs.Address
} });
const ValidatorCommissionCap = new import_utils.MoveStruct({ name: `${$moduleName}::ValidatorCommissionCap`, fields: {
  id: object.UID,
  validator_id: import_bcs.bcs.Address
} });
const VerifiedValidatorCap = new import_utils.MoveStruct({ name: `${$moduleName}::VerifiedValidatorCap`, fields: {
  validator_id: import_bcs.bcs.Address
} });
const VerifiedValidatorOperationCap = new import_utils.MoveStruct({ name: `${$moduleName}::VerifiedValidatorOperationCap`, fields: {
  validator_id: import_bcs.bcs.Address
} });
const VerifiedValidatorCommissionCap = new import_utils.MoveStruct({ name: `${$moduleName}::VerifiedValidatorCommissionCap`, fields: {
  validator_id: import_bcs.bcs.Address
} });
function newValidatorCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["validatorId", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "new_validator_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newValidatorOperationCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["validatorId", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "new_validator_operation_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function newValidatorCommissionCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["validatorId", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "new_validator_commission_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function createVerifiedValidatorCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["cap", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "create_verified_validator_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function createVerifiedValidatorOperationCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorOperationCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["cap", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "create_verified_validator_operation_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function createVerifiedValidatorCommissionCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCommissionCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["cap", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "create_verified_validator_commission_cap",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorOperationCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorOperationCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "validator_operation_cap_validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorCommissionCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCommissionCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "validator_commission_cap_validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifiedValidatorCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::VerifiedValidatorCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "verified_validator_cap_validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifiedValidatorOperationCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::VerifiedValidatorOperationCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "verified_validator_operation_cap_validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifiedValidatorCommissionCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::VerifiedValidatorCommissionCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "verified_validator_commission_cap_validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
