// @ika.xyz/sdk/dist/cjs/generated/ika_dwallet_2pc_mpc/ika_dwallet_2pc_mpc_init.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ika_dwallet_2pc_mpc_init_exports = {};
__export(ika_dwallet_2pc_mpc_init_exports, {
  IKA_DWALLET_2PC_MPC_INIT: () => IKA_DWALLET_2PC_MPC_INIT,
  InitCap: () => InitCap,
  initialize: () => initialize
});
module.exports = __toCommonJS(ika_dwallet_2pc_mpc_init_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
var _package = __toESM(require("./deps/sui/package.js"));
const $moduleName = "@local-pkg/2pc-mpc::ika_dwallet_2pc_mpc_init";
const IKA_DWALLET_2PC_MPC_INIT = new import_utils.MoveStruct({ name: `${$moduleName}::IKA_DWALLET_2PC_MPC_INIT`, fields: {
  dummy_field: import_bcs.bcs.bool()
} });
const InitCap = new import_utils.MoveStruct({ name: `${$moduleName}::InitCap`, fields: {
  id: object.UID,
  publisher: _package.Publisher
} });
function initialize(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::ika_dwallet_2pc_mpc_init::InitCap`,
    `${packageAddress}::advance_epoch_approver::AdvanceEpochApprover`,
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`,
    `${packageAddress}::pricing::PricingInfo`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::vec_map::VecMap<u32, 0x0000000000000000000000000000000000000000000000000000000000000002::vec_map::VecMap<u32, vector<u32>>>",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String"
  ];
  const parameterNames = ["initCap", "advanceEpochApprover", "systemCurrentStatusInfo", "pricing", "supportedCurvesToSignatureAlgorithmsToHashSchemes", "dwalletCapImageUrl", "importedKeyDwalletCapImageUrl", "unverifiedPresignCapImageUrl", "verifiedPresignCapImageUrl", "unverifiedPartialUserSignatureCapImageUrl", "verifiedPartialUserSignatureCapImageUrl"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "ika_dwallet_2pc_mpc_init",
    function: "initialize",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
