// @ika.xyz/sdk/dist/cjs/generated/ika_dwallet_2pc_mpc/pricing_and_fee_manager.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pricing_and_fee_manager_exports = {};
__export(pricing_and_fee_manager_exports, {
  PricingAndFeeManager: () => PricingAndFeeManager
});
module.exports = __toCommonJS(pricing_and_fee_manager_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var pricing = __toESM(require("./pricing.js"));
var table = __toESM(require("./deps/sui/table.js"));
var balance = __toESM(require("./deps/sui/balance.js"));
const $moduleName = "@local-pkg/2pc-mpc::pricing_and_fee_manager";
const PricingAndFeeManager = new import_utils.MoveStruct({ name: `${$moduleName}::PricingAndFeeManager`, fields: {
  /** Pricing for the current epoch */
  current: pricing.PricingInfo,
  /** Default pricing configuration */
  default: pricing.PricingInfo,
  /** Validator votes for pricing (validator ID -> pricing vote) */
  validator_votes: table.Table,
  /** Pricing calculation votes - if set, must complete before epoch advance */
  pricing_calculation_votes: import_bcs.bcs.option(pricing.PricingInfoCalculationVotes),
  /** Gas fee reimbursement value for system calls */
  gas_fee_reimbursement_sui_system_call_value: import_bcs.bcs.u64(),
  /** SUI balance for gas fee reimbursement to fund network tx responses */
  gas_fee_reimbursement_sui_system_call_balance: balance.Balance,
  /** IKA fees charged for consensus validation */
  fee_charged_ika: balance.Balance
} });
