// @ika.xyz/sdk/dist/cjs/generated/ika_dwallet_2pc_mpc/pricing.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pricing_exports = {};
__export(pricing_exports, {
  PricingInfo: () => PricingInfo,
  PricingInfoCalculationVotes: () => PricingInfoCalculationVotes,
  PricingInfoKey: () => PricingInfoKey,
  PricingInfoValue: () => PricingInfoValue,
  empty: () => empty,
  feeIka: () => feeIka,
  gasFeeReimbursementSui: () => gasFeeReimbursementSui,
  gasFeeReimbursementSuiForSystemCalls: () => gasFeeReimbursementSuiForSystemCalls,
  insertOrUpdatePricing: () => insertOrUpdatePricing
});
module.exports = __toCommonJS(pricing_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var vec_map = __toESM(require("./deps/sui/vec_map.js"));
var bls_committee = __toESM(require("./deps/ika_common/bls_committee.js"));
const $moduleName = "@local-pkg/2pc-mpc::pricing";
const PricingInfoKey = new import_utils.MoveStruct({ name: `${$moduleName}::PricingInfoKey`, fields: {
  curve: import_bcs.bcs.u32(),
  signature_algorithm: import_bcs.bcs.option(import_bcs.bcs.u32()),
  protocol: import_bcs.bcs.u32()
} });
const PricingInfoValue = new import_utils.MoveStruct({ name: `${$moduleName}::PricingInfoValue`, fields: {
  fee_ika: import_bcs.bcs.u64(),
  gas_fee_reimbursement_sui: import_bcs.bcs.u64(),
  gas_fee_reimbursement_sui_for_system_calls: import_bcs.bcs.u64()
} });
const PricingInfo = new import_utils.MoveStruct({ name: `${$moduleName}::PricingInfo`, fields: {
  /**
     * The pricing for each curve and signature algorithm and protocol. The first key
     * is the curve, the second is the signature algorithm, the third is the protocol.
     */
  pricing_map: vec_map.VecMap(PricingInfoKey, PricingInfoValue)
} });
const PricingInfoCalculationVotes = new import_utils.MoveStruct({ name: `${$moduleName}::PricingInfoCalculationVotes`, fields: {
  bls_committee: bls_committee.BlsCommittee,
  default_pricing: PricingInfo,
  working_pricing: PricingInfo
} });
function empty(options = {}) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "empty"
  });
}
function insertOrUpdatePricing(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfo`,
    "u32",
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u32>",
    "u32",
    "u64",
    "u64",
    "u64"
  ];
  const parameterNames = ["self", "curve", "signatureAlgorithm", "protocol", "feeIka", "gasFeeReimbursementSui", "gasFeeReimbursementSuiForSystemCalls"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "insert_or_update_pricing",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function feeIka(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfoValue`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "fee_ika",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function gasFeeReimbursementSui(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfoValue`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "gas_fee_reimbursement_sui",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function gasFeeReimbursementSuiForSystemCalls(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfoValue`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "gas_fee_reimbursement_sui_for_system_calls",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
