// @ika.xyz/sdk/dist/cjs/generated/ika_dwallet_2pc_mpc/support_config.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var support_config_exports = {};
__export(support_config_exports, {
  GlobalPresignConfig: () => GlobalPresignConfig,
  SupportConfig: () => SupportConfig
});
module.exports = __toCommonJS(support_config_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var vec_map = __toESM(require("./deps/sui/vec_map.js"));
const $moduleName = "@local-pkg/2pc-mpc::support_config";
const SupportConfig = new import_utils.MoveStruct({ name: `${$moduleName}::SupportConfig`, fields: {
  /**
     * A nested map of supported curves to signature algorithms to hash schemes. e.g.
     * secp256k1 -> [(ecdsa -> [sha256, keccak256]), (schnorr -> [sha256])]
     */
  supported_curves_to_signature_algorithms_to_hash_schemes: vec_map.VecMap(import_bcs.bcs.u32(), vec_map.VecMap(import_bcs.bcs.u32(), import_bcs.bcs.vector(import_bcs.bcs.u32()))),
  /** List of paused curves in case of emergency (e.g. [secp256k1, ristretto]) */
  paused_curves: import_bcs.bcs.vector(import_bcs.bcs.u32()),
  /** List of paused signature algorithms in case of emergency (e.g. [ecdsa, schnorr]) */
  paused_signature_algorithms: import_bcs.bcs.vector(import_bcs.bcs.u32()),
  /** List of paused hash schemes in case of emergency (e.g. [sha256, keccak256]) */
  paused_hash_schemes: import_bcs.bcs.vector(import_bcs.bcs.u32()),
  /**
   * Signature algorithms that are allowed for global presign Deprecated: Use
   * GlobalPresignConfig instead.
   */
  signature_algorithms_allowed_global_presign: import_bcs.bcs.vector(import_bcs.bcs.u32())
} });
const GlobalPresignConfig = new import_utils.MoveStruct({ name: `${$moduleName}::GlobalPresignConfig`, fields: {
  /**
     * Map of curves to signature algorithms for dWallets that are created via DKG.
     * This means for this curve and this signature algorithm, it is only allowed to
     * use global presign. e.g. secp256k1 -> [(ecdsa -> [sha256, keccak256]), (schnorr
     * -> [sha256])]
     */
  curve_to_signature_algorithms_for_dkg: vec_map.VecMap(import_bcs.bcs.u32(), import_bcs.bcs.vector(import_bcs.bcs.u32())),
  /**
   * Map of curves to signature algorithms for dWallets that are created via imported
   * key. This means for this curve and this signature algorithm, it is only allowed
   * to use global presign. e.g. secp256k1 -> [(ecdsa -> [sha256, keccak256]),
   * (schnorr -> [sha256])]
   */
  curve_to_signature_algorithms_for_imported_key: vec_map.VecMap(import_bcs.bcs.u32(), import_bcs.bcs.vector(import_bcs.bcs.u32()))
} });
