// @ika.xyz/sdk/dist/esm/generated/ika_common/bls_committee.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as group_ops from "./deps/sui/group_ops.js";
const $moduleName = "@local-pkg/common::bls_committee";
const BlsCommitteeMember = new MoveStruct({ name: `${$moduleName}::BlsCommitteeMember`, fields: {
  validator_id: bcs.Address,
  protocol_pubkey: group_ops.Element
} });
const BlsCommittee = new MoveStruct({ name: `${$moduleName}::BlsCommittee`, fields: {
  members: bcs.vector(BlsCommitteeMember),
  /** The aggregation of public keys for all members of the committee */
  aggregated_protocol_pubkey: group_ops.Element,
  /** Minimum signatures required for quorum (2n/3 + 1) */
  quorum_threshold: bcs.u64(),
  /** Minimum signatures required for validity (n/3 + 1) */
  validity_threshold: bcs.u64()
} });
const CommitteeQuorumVerifiedEvent = new MoveStruct({ name: `${$moduleName}::CommitteeQuorumVerifiedEvent`, fields: {
  epoch: bcs.u64(),
  signer_count: bcs.u64()
} });
function totalVotingPower(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "total_voting_power",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function quorumThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "quorum_threshold",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function validityThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "validity_threshold",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function newBlsCommitteeMember(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "0x0000000000000000000000000000000000000000000000000000000000000002::group_ops::Element<0x0000000000000000000000000000000000000000000000000000000000000002::bls12381::UncompressedG1>"
  ];
  const parameterNames = ["validatorId", "protocolPubkey"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "new_bls_committee_member",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommitteeMember`
  ];
  const parameterNames = ["member"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function newBlsCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `vector<${packageAddress}::bls_committee::BlsCommitteeMember>`
  ];
  const parameterNames = ["members"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "new_bls_committee",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function empty(options = {}) {
  const packageAddress = options.package ?? "@local-pkg/common";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "empty"
  });
}
function members(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "members",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorIds(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "validator_ids",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function contains(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "validatorId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "contains",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifyCertificate(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "u64",
    "vector<u8>",
    "vector<u8>",
    "vector<u8>"
  ];
  const parameterNames = ["self", "epoch", "signature", "signersBitmap", "intentBytes"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "verify_certificate",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isQuorumThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "u64"
  ];
  const parameterNames = ["self", "signerCount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "is_quorum_threshold",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isValidityThreshold(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::bls_committee::BlsCommittee`,
    "u64"
  ];
  const parameterNames = ["self", "signerCount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "bls_committee",
    function: "is_validity_threshold",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  BlsCommittee,
  BlsCommitteeMember,
  CommitteeQuorumVerifiedEvent,
  contains,
  empty,
  isQuorumThreshold,
  isValidityThreshold,
  members,
  newBlsCommittee,
  newBlsCommitteeMember,
  quorumThreshold,
  totalVotingPower,
  validatorId,
  validatorIds,
  validityThreshold,
  verifyCertificate
};
