// @ika.xyz/sdk/dist/esm/generated/ika_common/protocol_cap.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
const $moduleName = "@local-pkg/common::protocol_cap";
const ProtocolCap = new MoveStruct({ name: `${$moduleName}::ProtocolCap`, fields: {
  id: object.UID
} });
const VerifiedProtocolCap = new MoveStruct({ name: `${$moduleName}::VerifiedProtocolCap`, fields: {
  dummy_field: bcs.bool()
} });
function create(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "protocol_cap",
    function: "create",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function createVerified(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "protocol_cap",
    function: "create_verified",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  ProtocolCap,
  VerifiedProtocolCap,
  create,
  createVerified
};
