// @ika.xyz/sdk/dist/esm/generated/ika_common/system_current_status_info.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as bls_committee from "./bls_committee.js";
const $moduleName = "@local-pkg/common::system_current_status_info";
const SystemCurrentStatusInfo = new MoveStruct({ name: `${$moduleName}::SystemCurrentStatusInfo`, fields: {
  current_epoch: bcs.u64(),
  is_mid_epoch_time: bcs.bool(),
  is_end_epoch_time: bcs.bool(),
  current_epoch_active_committee: bls_committee.BlsCommittee,
  next_epoch_active_committee: bcs.option(bls_committee.BlsCommittee)
} });
function create(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "u64",
    "bool",
    "bool",
    `${packageAddress}::bls_committee::BlsCommittee`,
    `0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<${packageAddress}::bls_committee::BlsCommittee>`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["currentEpoch", "isMidEpochTime", "isEndEpochTime", "currentEpochActiveCommittee", "nextEpochActiveCommittee", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "create",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function currentEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "current_epoch",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isMidEpochTime(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "is_mid_epoch_time",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isEndEpochTime(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "is_end_epoch_time",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function currentEpochActiveCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "current_epoch_active_committee",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function nextEpochActiveCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "system_current_status_info",
    function: "next_epoch_active_committee",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  SystemCurrentStatusInfo,
  create,
  currentEpoch,
  currentEpochActiveCommittee,
  isEndEpochTime,
  isMidEpochTime,
  nextEpochActiveCommittee
};
