// @ika.xyz/sdk/dist/esm/generated/ika_common/upgrade_package_approver.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
const $moduleName = "@local-pkg/common::upgrade_package_approver";
const UpgradePackageApprover = new MoveStruct({ name: `${$moduleName}::UpgradePackageApprover`, fields: {
  upgrade_cap_id: bcs.Address,
  remaining_witnesses_to_approve: bcs.vector(bcs.string()),
  old_package_id: bcs.Address,
  new_package_id: bcs.option(bcs.Address),
  migration_epoch: bcs.u64()
} });
function create(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "vector<0x0000000000000000000000000000000000000000000000000000000000000001::string::String>",
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u64",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["upgradeCapId", "remainingWitnessesToApprove", "oldPackageId", "migrationEpoch", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "create",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function assertAllWitnessesApproved(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "assert_all_witnesses_approved",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function newPackageId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "new_package_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function oldPackageId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "old_package_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function migrationEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "migration_epoch",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function destroy(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["self", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "destroy",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function commit(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::package::UpgradeReceipt",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["self", "receipt", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "commit",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function approveUpgradePackageByWitness(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`,
    `${options.typeArguments[0]}`
  ];
  const parameterNames = ["upgradePackageApprover", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "upgrade_package_approver",
    function: "approve_upgrade_package_by_witness",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames),
    typeArguments: options.typeArguments
  });
}
export {
  UpgradePackageApprover,
  approveUpgradePackageByWitness,
  assertAllWitnessesApproved,
  commit,
  create,
  destroy,
  migrationEpoch,
  newPackageId,
  oldPackageId
};
