// @ika.xyz/sdk/dist/esm/generated/ika_common/validator_cap.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
const $moduleName = "@local-pkg/common::validator_cap";
const ValidatorCap = new MoveStruct({ name: `${$moduleName}::ValidatorCap`, fields: {
  id: object.UID,
  validator_id: bcs.Address
} });
const ValidatorOperationCap = new MoveStruct({ name: `${$moduleName}::ValidatorOperationCap`, fields: {
  id: object.UID,
  validator_id: bcs.Address
} });
const ValidatorCommissionCap = new MoveStruct({ name: `${$moduleName}::ValidatorCommissionCap`, fields: {
  id: object.UID,
  validator_id: bcs.Address
} });
const VerifiedValidatorCap = new MoveStruct({ name: `${$moduleName}::VerifiedValidatorCap`, fields: {
  validator_id: bcs.Address
} });
const VerifiedValidatorOperationCap = new MoveStruct({ name: `${$moduleName}::VerifiedValidatorOperationCap`, fields: {
  validator_id: bcs.Address
} });
const VerifiedValidatorCommissionCap = new MoveStruct({ name: `${$moduleName}::VerifiedValidatorCommissionCap`, fields: {
  validator_id: bcs.Address
} });
function newValidatorCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["validatorId", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "new_validator_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function newValidatorOperationCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["validatorId", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "new_validator_operation_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function newValidatorCommissionCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["validatorId", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "new_validator_commission_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function createVerifiedValidatorCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["cap", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "create_verified_validator_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function createVerifiedValidatorOperationCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorOperationCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["cap", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "create_verified_validator_operation_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function createVerifiedValidatorCommissionCap(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCommissionCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`
  ];
  const parameterNames = ["cap", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "create_verified_validator_commission_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorOperationCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorOperationCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "validator_operation_cap_validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorCommissionCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::ValidatorCommissionCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "validator_commission_cap_validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifiedValidatorCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::VerifiedValidatorCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "verified_validator_cap_validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifiedValidatorOperationCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::VerifiedValidatorOperationCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "verified_validator_operation_cap_validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifiedValidatorCommissionCapValidatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/common";
  const argumentsTypes = [
    `${packageAddress}::validator_cap::VerifiedValidatorCommissionCap`
  ];
  const parameterNames = ["cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_cap",
    function: "verified_validator_commission_cap_validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  ValidatorCap,
  ValidatorCommissionCap,
  ValidatorOperationCap,
  VerifiedValidatorCap,
  VerifiedValidatorCommissionCap,
  VerifiedValidatorOperationCap,
  createVerifiedValidatorCap,
  createVerifiedValidatorCommissionCap,
  createVerifiedValidatorOperationCap,
  newValidatorCap,
  newValidatorCommissionCap,
  newValidatorOperationCap,
  validatorCommissionCapValidatorId,
  validatorId,
  validatorOperationCapValidatorId,
  verifiedValidatorCapValidatorId,
  verifiedValidatorCommissionCapValidatorId,
  verifiedValidatorOperationCapValidatorId
};
