// @ika.xyz/sdk/dist/esm/generated/ika_system/init.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
import * as _package from "./deps/sui/package.js";
const $moduleName = "@local-pkg/system::init";
const INIT = new MoveStruct({ name: `${$moduleName}::INIT`, fields: {
  dummy_field: bcs.bool()
} });
const InitCap = new MoveStruct({ name: `${$moduleName}::InitCap`, fields: {
  id: object.UID,
  publisher: _package.Publisher
} });
function initialize(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::init::InitCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::package::UpgradeCap",
    "0x0000000000000000000000000000000000000000000000000000000000000002::package::UpgradeCap",
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::TreasuryCap<${packageAddress}::ika::IKA>`,
    "u64",
    "u64",
    "u64",
    "u64",
    "u16",
    "u64",
    "u64",
    "u64",
    "u64",
    "u16",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String"
  ];
  const parameterNames = ["initCap", "systemObjectCap", "ikaUpgradeCap", "ikaSystemUpgradeCap", "protocolTreasuryCap", "protocolVersion", "chainStartTimestampMs", "epochDurationMs", "stakeSubsidyStartEpoch", "stakeSubsidyRate", "stakeSubsidyPeriodLength", "minValidatorCount", "maxValidatorCount", "minValidatorJoiningStake", "rewardSlashingRate", "stakedIkaImageUrl"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "init",
    function: "initialize",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  INIT,
  InitCap,
  initialize
};
