// @ika.xyz/sdk/dist/esm/generated/ika_system/staked_ika.js
import { MoveEnum, MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
import * as balance from "./deps/sui/balance.js";
const $moduleName = "@local-pkg/system::staked_ika";
const StakedIkaState = new MoveEnum({ name: `${$moduleName}::StakedIkaState`, fields: {
  Staked: null,
  Withdrawing: new MoveStruct({ name: `StakedIkaState.Withdrawing`, fields: {
    withdraw_epoch: bcs.u64()
  } })
} });
const StakedIka = new MoveStruct({ name: `${$moduleName}::StakedIka`, fields: {
  id: object.UID,
  /** Whether the staked IKA is active or withdrawing. */
  state: StakedIkaState,
  /** ID of the staking pool. */
  validator_id: bcs.Address,
  /** The staked amount. */
  principal: balance.Balance,
  /** The Ikarus epoch when the staked IKA was activated. */
  activation_epoch: bcs.u64()
} });
function validatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "validator_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function value(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "value",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function activationEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "activation_epoch",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isStaked(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "is_staked",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isWithdrawing(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "is_withdrawing",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function withdrawEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "withdraw_epoch",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function join(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`,
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw", "other"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "join",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function split(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`,
    "u64"
  ];
  const parameterNames = ["sw", "amount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "split",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  StakedIka,
  StakedIkaState,
  activationEpoch,
  isStaked,
  isWithdrawing,
  join,
  split,
  validatorId,
  value,
  withdrawEpoch
};
