// @ika.xyz/ika-wasm/dist/bundler/dwallet_mpc_wasm_bg.js
let wasm;
export function __wbg_set_wasm(val) {
    wasm = val;
}


let WASM_VECTOR_LEN = 0;

let cachedUint8ArrayMemory0 = null;

function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

const lTextEncoder = typeof TextEncoder === 'undefined' ? (0, module.require)('util').TextEncoder : TextEncoder;

let cachedTextEncoder = new lTextEncoder('utf-8');

const encodeString = (typeof cachedTextEncoder.encodeInto === 'function'
    ? function (arg, view) {
    return cachedTextEncoder.encodeInto(arg, view);
}
    : function (arg, view) {
    const buf = cachedTextEncoder.encode(arg);
    view.set(buf);
    return {
        read: arg.length,
        written: buf.length
    };
});

function passStringToWasm0(arg, malloc, realloc) {

    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;

    const mem = getUint8ArrayMemory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }

    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = encodeString(arg, view);

        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

let cachedDataViewMemory0 = null;

function getDataViewMemory0() {
    if (cachedDataViewMemory0 === null || cachedDataViewMemory0.buffer.detached === true || (cachedDataViewMemory0.buffer.detached === undefined && cachedDataViewMemory0.buffer !== wasm.memory.buffer)) {
        cachedDataViewMemory0 = new DataView(wasm.memory.buffer);
    }
    return cachedDataViewMemory0;
}

function getArrayU8FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint8ArrayMemory0().subarray(ptr / 1, ptr / 1 + len);
}

function addToExternrefTable0(obj) {
    const idx = wasm.__externref_table_alloc();
    wasm.__wbindgen_export_4.set(idx, obj);
    return idx;
}

function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        const idx = addToExternrefTable0(e);
        wasm.__wbindgen_exn_store(idx);
    }
}

const lTextDecoder = typeof TextDecoder === 'undefined' ? (0, module.require)('util').TextDecoder : TextDecoder;

let cachedTextDecoder = new lTextDecoder('utf-8', { ignoreBOM: true, fatal: true });

cachedTextDecoder.decode();

function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}

function passArray8ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 1, 1) >>> 0;
    getUint8ArrayMemory0().set(arg, ptr / 1);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function takeFromExternrefTable0(idx) {
    const value = wasm.__wbindgen_export_4.get(idx);
    wasm.__externref_table_dealloc(idx);
    return value;
}
/**
 * @param {Uint8Array} protocol_pp
 * @param {Uint8Array} decentralized_first_round_public_output
 * @returns {any}
 */
export function create_dkg_centralized_output_v1(protocol_pp, decentralized_first_round_public_output) {
    const ptr0 = passArray8ToWasm0(protocol_pp, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(decentralized_first_round_public_output, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.create_dkg_centralized_output_v1(ptr0, len0, ptr1, len1);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {Uint8Array} protocol_pp
 * @param {Uint8Array} session_identifier
 * @returns {any}
 */
export function create_dkg_centralized_output_v2(curve, protocol_pp, session_identifier) {
    const ptr0 = passArray8ToWasm0(protocol_pp, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(session_identifier, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.create_dkg_centralized_output_v2(curve, ptr0, len0, ptr1, len1);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {Uint8Array} dwallet_output
 * @returns {any}
 */
export function public_key_from_dwallet_output(curve, dwallet_output) {
    const ptr0 = passArray8ToWasm0(dwallet_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.public_key_from_dwallet_output(curve, ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {Uint8Array} centralized_dkg_output
 * @returns {any}
 */
export function public_key_from_centralized_dkg_output(curve, centralized_dkg_output) {
    const ptr0 = passArray8ToWasm0(centralized_dkg_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.public_key_from_centralized_dkg_output(curve, ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {Uint8Array} network_key_bytes
 * @returns {any}
 */
export function network_key_version(network_key_bytes) {
    const ptr0 = passArray8ToWasm0(network_key_bytes, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.network_key_version(ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {Uint8Array} dwallet_output_bytes
 * @returns {any}
 */
export function dwallet_version(dwallet_output_bytes) {
    const ptr0 = passArray8ToWasm0(dwallet_output_bytes, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.dwallet_version(ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * Derives a class groups keypair from a given seed.
 *
 * The class groups public encryption key being used to encrypt a keypair will be
 * different from the encryption key used to encrypt a Ristretto keypair.
 * The plaintext space/fundamental group will correspond to the order
 * of the respective elliptic curve.
 * The secret decryption key may be the same in terms of correctness,
 * but to simplify security analysis and implementation current version maintain distinct key-pairs.
 * @param {number} curve
 * @param {Uint8Array} seed
 * @returns {any}
 */
export function generate_secp_cg_keypair_from_seed(curve, seed) {
    const ptr0 = passArray8ToWasm0(seed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.generate_secp_cg_keypair_from_seed(curve, ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {Uint8Array} network_dkg_public_output
 * @returns {any}
 */
export function network_dkg_public_output_to_protocol_pp(curve, network_dkg_public_output) {
    const ptr0 = passArray8ToWasm0(network_dkg_public_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.network_dkg_public_output_to_protocol_pp(curve, ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {Uint8Array} reconfig_public_output
 * @param {Uint8Array} network_dkg_public_output
 * @returns {any}
 */
export function reconfiguration_public_output_to_protocol_pp(curve, reconfig_public_output, network_dkg_public_output) {
    const ptr0 = passArray8ToWasm0(reconfig_public_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(network_dkg_public_output, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.reconfiguration_public_output_to_protocol_pp(curve, ptr0, len0, ptr1, len1);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {Uint8Array} centralized_dkg_output
 * @param {Uint8Array} decentralized_dkg_output
 * @returns {any}
 */
export function centralized_and_decentralized_parties_dkg_output_match(curve, centralized_dkg_output, decentralized_dkg_output) {
    const ptr0 = passArray8ToWasm0(centralized_dkg_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(decentralized_dkg_output, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.centralized_and_decentralized_parties_dkg_output_match(curve, ptr0, len0, ptr1, len1);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * Encrypts the given secret share to the given encryption key.
 * Returns a tuple of the encryption key and proof of encryption.
 * @param {number} curve
 * @param {Uint8Array} secret_key_share
 * @param {Uint8Array} encryption_key
 * @param {Uint8Array} protocol_pp
 * @returns {any}
 */
export function encrypt_secret_share(curve, secret_key_share, encryption_key, protocol_pp) {
    const ptr0 = passArray8ToWasm0(secret_key_share, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(encryption_key, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(protocol_pp, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ret = wasm.encrypt_secret_share(curve, ptr0, len0, ptr1, len1, ptr2, len2);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * Decrypts the given encrypted user share using the given decryption key.
 * @param {number} curve
 * @param {Uint8Array} decryption_key
 * @param {Uint8Array} dwallet_dkg_output
 * @param {Uint8Array} encrypted_user_share_and_proof
 * @param {Uint8Array} protocol_pp
 * @returns {any}
 */
export function decrypt_user_share(curve, decryption_key, dwallet_dkg_output, encrypted_user_share_and_proof, protocol_pp) {
    const ptr0 = passArray8ToWasm0(decryption_key, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(dwallet_dkg_output, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(encrypted_user_share_and_proof, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ptr3 = passArray8ToWasm0(protocol_pp, wasm.__wbindgen_malloc);
    const len3 = WASM_VECTOR_LEN;
    const ret = wasm.decrypt_user_share(curve, ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * Verifies that the given secret key share matches the given dWallet public key share.
 * DKG output->centralized_party_public_key_share.
 * @param {number} curve
 * @param {Uint8Array} secret_share
 * @param {Uint8Array} dkg_output
 * @param {Uint8Array} network_dkg_public_output
 * @returns {any}
 */
export function verify_user_share(curve, secret_share, dkg_output, network_dkg_public_output) {
    const ptr0 = passArray8ToWasm0(secret_share, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(dkg_output, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(network_dkg_public_output, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ret = wasm.verify_user_share(curve, ptr0, len0, ptr1, len1, ptr2, len2);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {Uint8Array} network_dkg_public_output
 * @returns {any}
 */
export function sample_dwallet_keypair(network_dkg_public_output) {
    const ptr0 = passArray8ToWasm0(network_dkg_public_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.sample_dwallet_keypair(ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {Uint8Array} public_key
 * @param {Uint8Array} signature
 * @param {Uint8Array} message
 * @param {Uint8Array} network_dkg_public_output
 * @param {number} curve
 * @param {number} signature_algorithm
 * @param {number} hash_scheme
 * @returns {any}
 */
export function verify_secp_signature(public_key, signature, message, network_dkg_public_output, curve, signature_algorithm, hash_scheme) {
    const ptr0 = passArray8ToWasm0(public_key, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(signature, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(message, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ptr3 = passArray8ToWasm0(network_dkg_public_output, wasm.__wbindgen_malloc);
    const len3 = WASM_VECTOR_LEN;
    const ret = wasm.verify_secp_signature(ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3, curve, signature_algorithm, hash_scheme);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {Uint8Array} network_dkg_public_output
 * @param {Uint8Array} session_identifier
 * @param {Uint8Array} secret_share
 * @returns {any}
 */
export function create_imported_dwallet_centralized_step(curve, network_dkg_public_output, session_identifier, secret_share) {
    const ptr0 = passArray8ToWasm0(network_dkg_public_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(session_identifier, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(secret_share, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ret = wasm.create_imported_dwallet_centralized_step(curve, ptr0, len0, ptr1, len1, ptr2, len2);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {Uint8Array} protocol_pp
 * @param {Uint8Array} decentralized_party_dkg_public_output
 * @param {Uint8Array} centralized_party_dkg_secret_output
 * @param {Uint8Array} presign
 * @param {Uint8Array} message
 * @param {number} curve
 * @param {number} signature_algorithm
 * @param {number} hash_scheme
 * @returns {any}
 */
export function create_sign_centralized_party_message(protocol_pp, decentralized_party_dkg_public_output, centralized_party_dkg_secret_output, presign, message, curve, signature_algorithm, hash_scheme) {
    const ptr0 = passArray8ToWasm0(protocol_pp, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(decentralized_party_dkg_public_output, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(centralized_party_dkg_secret_output, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ptr3 = passArray8ToWasm0(presign, wasm.__wbindgen_malloc);
    const len3 = WASM_VECTOR_LEN;
    const ptr4 = passArray8ToWasm0(message, wasm.__wbindgen_malloc);
    const len4 = WASM_VECTOR_LEN;
    const ret = wasm.create_sign_centralized_party_message(ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3, ptr4, len4, curve, signature_algorithm, hash_scheme);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {Uint8Array} protocol_pp
 * @param {Uint8Array} centralized_party_dkg_public_output
 * @param {Uint8Array} centralized_party_dkg_secret_output
 * @param {Uint8Array} presign
 * @param {Uint8Array} message
 * @param {number} hash_scheme
 * @param {number} signature_algorithm
 * @param {number} curve
 * @returns {any}
 */
export function create_sign_centralized_party_message_with_centralized_party_dkg_output(protocol_pp, centralized_party_dkg_public_output, centralized_party_dkg_secret_output, presign, message, hash_scheme, signature_algorithm, curve) {
    const ptr0 = passArray8ToWasm0(protocol_pp, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(centralized_party_dkg_public_output, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArray8ToWasm0(centralized_party_dkg_secret_output, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ptr3 = passArray8ToWasm0(presign, wasm.__wbindgen_malloc);
    const len3 = WASM_VECTOR_LEN;
    const ptr4 = passArray8ToWasm0(message, wasm.__wbindgen_malloc);
    const len4 = WASM_VECTOR_LEN;
    const ret = wasm.create_sign_centralized_party_message_with_centralized_party_dkg_output(ptr0, len0, ptr1, len1, ptr2, len2, ptr3, len3, ptr4, len4, hash_scheme, signature_algorithm, curve);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

/**
 * @param {number} curve
 * @param {number} signature_algorithm
 * @param {Uint8Array} signature_output
 * @returns {any}
 */
export function parse_signature_from_sign_output(curve, signature_algorithm, signature_output) {
    const ptr0 = passArray8ToWasm0(signature_output, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.parse_signature_from_sign_output(curve, signature_algorithm, ptr0, len0);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}

export function __wbg_String_8f0eb39a4a4c2f66(arg0, arg1) {
    const ret = String(arg1);
    const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
};

export function __wbg_getRandomValues_3c9c0d586e575a16() { return handleError(function (arg0, arg1) {
    globalThis.crypto.getRandomValues(getArrayU8FromWasm0(arg0, arg1));
}, arguments) };

export function __wbg_new_78feb108b6472713() {
    const ret = new Array();
    return ret;
};

export function __wbg_set_37837023f3d740e8(arg0, arg1, arg2) {
    arg0[arg1 >>> 0] = arg2;
};

export function __wbindgen_error_new(arg0, arg1) {
    const ret = new Error(getStringFromWasm0(arg0, arg1));
    return ret;
};

export function __wbindgen_init_externref_table() {
    const table = wasm.__wbindgen_export_4;
    const offset = table.grow(4);
    table.set(0, undefined);
    table.set(offset + 0, undefined);
    table.set(offset + 1, null);
    table.set(offset + 2, true);
    table.set(offset + 3, false);
    ;
};

export function __wbindgen_number_new(arg0) {
    const ret = arg0;
    return ret;
};

export function __wbindgen_throw(arg0, arg1) {
    throw new Error(getStringFromWasm0(arg0, arg1));
};
