// @ika.xyz/sdk/dist/cjs/generated/ika_system/init.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var init_exports = {};
__export(init_exports, {
  INIT: () => INIT,
  InitCap: () => InitCap,
  initialize: () => initialize
});
module.exports = __toCommonJS(init_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
var _package = __toESM(require("./deps/sui/package.js"));
const $moduleName = "@local-pkg/system::init";
const INIT = new import_utils.MoveStruct({ name: `${$moduleName}::INIT`, fields: {
  dummy_field: import_bcs.bcs.bool()
} });
const InitCap = new import_utils.MoveStruct({ name: `${$moduleName}::InitCap`, fields: {
  id: object.UID,
  publisher: _package.Publisher
} });
function initialize(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::init::InitCap`,
    `${packageAddress}::system_object_cap::SystemObjectCap`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::package::UpgradeCap",
    "0x0000000000000000000000000000000000000000000000000000000000000002::package::UpgradeCap",
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::TreasuryCap<${packageAddress}::ika::IKA>`,
    "u64",
    "u64",
    "u64",
    "u64",
    "u16",
    "u64",
    "u64",
    "u64",
    "u64",
    "u16",
    "0x0000000000000000000000000000000000000000000000000000000000000001::string::String"
  ];
  const parameterNames = ["initCap", "systemObjectCap", "ikaUpgradeCap", "ikaSystemUpgradeCap", "protocolTreasuryCap", "protocolVersion", "chainStartTimestampMs", "epochDurationMs", "stakeSubsidyStartEpoch", "stakeSubsidyRate", "stakeSubsidyPeriodLength", "minValidatorCount", "maxValidatorCount", "minValidatorJoiningStake", "rewardSlashingRate", "stakedIkaImageUrl"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "init",
    function: "initialize",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
