// @ika.xyz/sdk/dist/cjs/generated/ika_system/pending_active_set.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pending_active_set_exports = {};
__export(pending_active_set_exports, {
  PendingActiveSet: () => PendingActiveSet,
  PendingActiveSetEntry: () => PendingActiveSetEntry
});
module.exports = __toCommonJS(pending_active_set_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var vec_set = __toESM(require("./deps/sui/vec_set.js"));
const $moduleName = "@local-pkg/system::pending_active_set";
const PendingActiveSetEntry = new import_utils.MoveStruct({ name: `${$moduleName}::PendingActiveSetEntry`, fields: {
  /** The ID of the validator */
  validator_id: import_bcs.bcs.Address,
  /** The amount of IKA staked by this validator */
  staked_amount: import_bcs.bcs.u64()
} });
const PendingActiveSet = new import_utils.MoveStruct({ name: `${$moduleName}::PendingActiveSet`, fields: {
  /** The minimum number of validators required in the active set */
  min_validator_count: import_bcs.bcs.u64(),
  /** The maximum number of validators in the active set */
  max_validator_count: import_bcs.bcs.u64(),
  /**
   * The minimum amount of staked IKA needed to enter the active set. This is used to
   * determine if a storage validator can be added to the active set
   */
  min_validator_joining_stake: import_bcs.bcs.u64(),
  /**
   * The maximum number of validators that can be added or removed to the active set
   * in an epoch
   */
  max_validator_change_count: import_bcs.bcs.u64(),
  /** The list of validators in the active set and their stake */
  validators: import_bcs.bcs.vector(PendingActiveSetEntry),
  /** The total amount of staked IKA in the active set */
  total_stake: import_bcs.bcs.u64(),
  /**
   * The list of validators that have been added or removed to the active set in the
   * current epoch
   */
  validator_changes: vec_set.VecSet(import_bcs.bcs.Address)
} });
