// @ika.xyz/sdk/dist/cjs/generated/ika_system/protocol_treasury.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var protocol_treasury_exports = {};
__export(protocol_treasury_exports, {
  ProtocolTreasury: () => ProtocolTreasury,
  stakeSubsidyAmountPerDistribution: () => stakeSubsidyAmountPerDistribution
});
module.exports = __toCommonJS(protocol_treasury_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var coin = __toESM(require("./deps/sui/coin.js"));
var bag = __toESM(require("./deps/sui/bag.js"));
const $moduleName = "@local-pkg/system::protocol_treasury";
const ProtocolTreasury = new import_utils.MoveStruct({ name: `${$moduleName}::ProtocolTreasury`, fields: {
  /** TreasuryCap of IKA tokens. */
  treasury_cap: coin.TreasuryCap,
  /** Count of the number of times stake subsidies have been distributed. */
  stake_subsidy_distribution_counter: import_bcs.bcs.u64(),
  /**
   * The rate at which the amount per distribution is calculated based on period nad
   * total supply. Expressed in basis points.
   */
  stake_subsidy_rate: import_bcs.bcs.u16(),
  /**
   * The amount of stake subsidy to be destructured per distribution. This amount
   * changes based on `stake_subsidy_rate`.
   */
  stake_subsidy_amount_per_distribution: import_bcs.bcs.u64(),
  /**
   * Number of distributions to occur before the amount per distribution will be
   * recalculated.
   */
  stake_subsidy_period_length: import_bcs.bcs.u64(),
  /** The total supply of IKA tokens at the start of the current period. */
  total_supply_at_period_start: import_bcs.bcs.u64(),
  /** Any extra fields that's not defined statically. */
  extra_fields: bag.Bag
} });
function stakeSubsidyAmountPerDistribution(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::protocol_treasury::ProtocolTreasury`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "protocol_treasury",
    function: "stake_subsidy_amount_per_distribution",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
