// @ika.xyz/sdk/dist/cjs/generated/ika_system/staked_ika.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var staked_ika_exports = {};
__export(staked_ika_exports, {
  StakedIka: () => StakedIka,
  StakedIkaState: () => StakedIkaState,
  activationEpoch: () => activationEpoch,
  isStaked: () => isStaked,
  isWithdrawing: () => isWithdrawing,
  join: () => join,
  split: () => split,
  validatorId: () => validatorId,
  value: () => value,
  withdrawEpoch: () => withdrawEpoch
});
module.exports = __toCommonJS(staked_ika_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/sui/object.js"));
var balance = __toESM(require("./deps/sui/balance.js"));
const $moduleName = "@local-pkg/system::staked_ika";
const StakedIkaState = new import_utils.MoveEnum({ name: `${$moduleName}::StakedIkaState`, fields: {
  Staked: null,
  Withdrawing: new import_utils.MoveStruct({ name: `StakedIkaState.Withdrawing`, fields: {
    withdraw_epoch: import_bcs.bcs.u64()
  } })
} });
const StakedIka = new import_utils.MoveStruct({ name: `${$moduleName}::StakedIka`, fields: {
  id: object.UID,
  /** Whether the staked IKA is active or withdrawing. */
  state: StakedIkaState,
  /** ID of the staking pool. */
  validator_id: import_bcs.bcs.Address,
  /** The staked amount. */
  principal: balance.Balance,
  /** The Ikarus epoch when the staked IKA was activated. */
  activation_epoch: import_bcs.bcs.u64()
} });
function validatorId(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "validator_id",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function value(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "value",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function activationEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "activation_epoch",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isStaked(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "is_staked",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isWithdrawing(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "is_withdrawing",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function withdrawEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "withdraw_epoch",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function join(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`,
    `${packageAddress}::staked_ika::StakedIka`
  ];
  const parameterNames = ["sw", "other"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "join",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function split(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::staked_ika::StakedIka`,
    "u64"
  ];
  const parameterNames = ["sw", "amount"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "staked_ika",
    function: "split",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
