// @ika.xyz/sdk/dist/cjs/generated/ika_system/validator_set.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validator_set_exports = {};
__export(validator_set_exports, {
  ValidatorEpochInfoEventV1: () => ValidatorEpochInfoEventV1,
  ValidatorJoinEvent: () => ValidatorJoinEvent,
  ValidatorLeaveEvent: () => ValidatorLeaveEvent,
  ValidatorSet: () => ValidatorSet,
  activeCommittee: () => activeCommittee,
  getValidator: () => getValidator,
  isInactiveValidator: () => isInactiveValidator,
  isValidatorCandidate: () => isValidatorCandidate,
  nextEpochActiveCommittee: () => nextEpochActiveCommittee,
  pendingActiveSet: () => pendingActiveSet,
  totalStake: () => totalStake,
  validatorTotalStakeAmount: () => validatorTotalStakeAmount
});
module.exports = __toCommonJS(validator_set_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var object_table = __toESM(require("./deps/sui/object_table.js"));
var bls_committee = __toESM(require("./deps/ika_common/bls_committee.js"));
var extended_field = __toESM(require("./deps/ika_common/extended_field.js"));
var vec_map = __toESM(require("./deps/sui/vec_map.js"));
var vec_set = __toESM(require("./deps/sui/vec_set.js"));
var bag = __toESM(require("./deps/sui/bag.js"));
var token_exchange_rate = __toESM(require("./token_exchange_rate.js"));
const $moduleName = "@local-pkg/system::validator_set";
const ValidatorSet = new import_utils.MoveStruct({ name: `${$moduleName}::ValidatorSet`, fields: {
  /** Total amount of stake from all active validators at the beginning of the epoch. */
  total_stake: import_bcs.bcs.u64(),
  /** How many reward are slashed to punish a validator, in bps. */
  reward_slashing_rate: import_bcs.bcs.u16(),
  /** A table that contains all validators */
  validators: object_table.ObjectTable,
  /** The current list of active committee of validators. */
  active_committee: bls_committee.BlsCommittee,
  /**
   * The next list of active committee of validators. It will become the
   * active_committee at the end of the epoch.
   */
  next_epoch_active_committee: import_bcs.bcs.option(bls_committee.BlsCommittee),
  /** The current list of previous committee of validators. */
  previous_committee: bls_committee.BlsCommittee,
  /**
   * The next list of pending active set of validators to be
   * next_epoch_active_committee. It will start from the last
   * next_epoch_active_committee and will be process between middle of the epochs and
   * will be finalize at the middle of the epoch.
   */
  pending_active_set: extended_field.ExtendedField,
  /**
   * A map storing the records of validator reporting each other. There is an entry
   * in the map for each validator that has been reported at least once. The entry
   * VecSet contains all the validators that reported them. If a validator has never
   * been reported they don't have an entry in this map. This map persists across
   * epoch: a peer continues being in a reported state until the reporter doesn't
   * explicitly remove their report. Note that in case we want to support validator
   * address change in future, the reports should be based on validator ids
   */
  validator_report_records: vec_map.VecMap(import_bcs.bcs.Address, vec_set.VecSet(import_bcs.bcs.Address)),
  /** Any extra fields that's not defined statically. */
  extra_fields: bag.Bag
} });
const ValidatorEpochInfoEventV1 = new import_utils.MoveStruct({ name: `${$moduleName}::ValidatorEpochInfoEventV1`, fields: {
  epoch: import_bcs.bcs.u64(),
  validator_id: import_bcs.bcs.Address,
  stake: import_bcs.bcs.u64(),
  commission_rate: import_bcs.bcs.u16(),
  staking_rewards: import_bcs.bcs.u64(),
  token_exchange_rate: token_exchange_rate.TokenExchangeRate,
  tallying_rule_reporters: import_bcs.bcs.vector(import_bcs.bcs.Address),
  tallying_rule_global_score: import_bcs.bcs.u64()
} });
const ValidatorJoinEvent = new import_utils.MoveStruct({ name: `${$moduleName}::ValidatorJoinEvent`, fields: {
  epoch: import_bcs.bcs.u64(),
  validator_id: import_bcs.bcs.Address
} });
const ValidatorLeaveEvent = new import_utils.MoveStruct({ name: `${$moduleName}::ValidatorLeaveEvent`, fields: {
  withdrawing_epoch: import_bcs.bcs.u64(),
  validator_id: import_bcs.bcs.Address,
  is_voluntary: import_bcs.bcs.bool()
} });
function totalStake(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "total_stake",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function validatorTotalStakeAmount(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "validatorId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "validator_total_stake_amount",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function getValidator(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "validatorId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "get_validator",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function activeCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "active_committee",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function nextEpochActiveCommittee(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "next_epoch_active_committee",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function pendingActiveSet(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "pending_active_set",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isValidatorCandidate(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "validatorId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "is_validator_candidate",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function isInactiveValidator(options) {
  const packageAddress = options.package ?? "@local-pkg/system";
  const argumentsTypes = [
    `${packageAddress}::validator_set::ValidatorSet`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "validatorId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "validator_set",
    function: "is_inactive_validator",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
