// @ika.xyz/sdk/dist/cjs/generated/ika_dwallet_2pc_mpc/deps/ika_common/bls_committee.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bls_committee_exports = {};
__export(bls_committee_exports, {
  BlsCommittee: () => BlsCommittee,
  BlsCommitteeMember: () => BlsCommitteeMember
});
module.exports = __toCommonJS(bls_committee_exports);
var import_utils = require("../../../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
var group_ops = __toESM(require("../sui/group_ops.js"));
const $moduleName = "ika_common::bls_committee";
const BlsCommitteeMember = new import_utils.MoveStruct({ name: `${$moduleName}::BlsCommitteeMember`, fields: {
  validator_id: import_bcs.bcs.Address,
  protocol_pubkey: group_ops.Element
} });
const BlsCommittee = new import_utils.MoveStruct({ name: `${$moduleName}::BlsCommittee`, fields: {
  members: import_bcs.bcs.vector(BlsCommitteeMember),
  /** The aggregation of public keys for all members of the committee */
  aggregated_protocol_pubkey: group_ops.Element,
  /** Minimum signatures required for quorum (2n/3 + 1) */
  quorum_threshold: import_bcs.bcs.u64(),
  /** Minimum signatures required for validity (n/3 + 1) */
  validity_threshold: import_bcs.bcs.u64()
} });
