// @ika.xyz/sdk/dist/esm/client/errors.js
class IkaClientError extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "IkaClientError";
  }
}
class ObjectNotFoundError extends IkaClientError {
  constructor(objectType, objectId, cause) {
    const idInfo = objectId ? ` with ID ${objectId}` : "";
    super(`${objectType} object${idInfo} not found`, cause);
    this.name = "ObjectNotFoundError";
  }
}
class InvalidObjectError extends IkaClientError {
  constructor(objectType, objectId, cause) {
    const idInfo = objectId ? ` (ID: ${objectId})` : "";
    super(`Invalid ${objectType} object${idInfo}: Expected structure not found`, cause);
    this.name = "InvalidObjectError";
  }
}
class CacheError extends IkaClientError {
  constructor(message, cause) {
    super(`Cache error: ${message}`, cause);
    this.name = "CacheError";
  }
}
class NetworkError extends IkaClientError {
  constructor(message, cause) {
    super(`Network error: ${message}`, cause);
    this.name = "NetworkError";
  }
}
export {
  CacheError,
  IkaClientError,
  InvalidObjectError,
  NetworkError,
  ObjectNotFoundError
};
