// @ika.xyz/sdk/dist/esm/client/types.js
import { bcs, BcsStruct } from "@mysten/sui/bcs";
import * as CoordinatorInnerModule from "../generated/ika_dwallet_2pc_mpc/coordinator_inner.js";
import * as SystemInnerModule from "../generated/ika_system/system_inner.js";
const DWalletKind = {
  ZeroTrust: "zero-trust",
  ImportedKey: "imported-key",
  ImportedKeyShared: "imported-key-shared",
  Shared: "shared"
};
const Hash = {
  /** KECCAK256 (SHA3) - Compatible with: ECDSASecp256k1 */
  KECCAK256: "KECCAK256",
  /** SHA256 - Compatible with: ECDSASecp256k1, Taproot, ECDSASecp256r1 */
  SHA256: "SHA256",
  /** Double SHA256: h(x) = sha256(sha256(x)) - Compatible with: ECDSASecp256k1, ECDSASecp256r1 */
  DoubleSHA256: "DoubleSHA256",
  /** SHA512 - Compatible with: EdDSA only */
  SHA512: "SHA512",
  /** Merlin (STROBE-based transcript construction) - Compatible with: SchnorrkelSubstrate only */
  Merlin: "Merlin"
};
const Curve = {
  /** secp256k1 - Used by: ECDSASecp256k1, Taproot */
  SECP256K1: "SECP256K1",
  /** Ristretto - Used by: SchnorrkelSubstrate */
  RISTRETTO: "RISTRETTO",
  /** Ed25519 - Used by: EdDSA */
  ED25519: "ED25519",
  /** secp256r1 (P-256) - Used by: ECDSASecp256r1 */
  SECP256R1: "SECP256R1"
};
const SignatureAlgorithm = {
  /** ECDSA with secp256k1 curve - Valid hashes: KECCAK256, SHA256, DoubleSHA256 */
  ECDSASecp256k1: "ECDSASecp256k1",
  /** Taproot (Bitcoin) - Valid hash: SHA256 only */
  Taproot: "Taproot",
  /** ECDSA with secp256r1 (P-256) curve - Valid hashes: SHA256, DoubleSHA256 */
  ECDSASecp256r1: "ECDSASecp256r1",
  /** EdDSA (Ed25519) - Valid hash: SHA512 only */
  EdDSA: "EdDSA",
  /** Schnorrkel/Ristretto (Substrate) - Valid hash: Merlin only */
  SchnorrkelSubstrate: "SchnorrkelSubstrate"
};
function DynamicField(...typeParameters) {
  return new BcsStruct({
    name: `dynamic_field::Field<u64, ${typeParameters[0].name}>`,
    fields: {
      id: bcs.Address,
      name: bcs.u64(),
      value: typeParameters[0]
    }
  });
}
const CoordinatorInnerDynamicField = DynamicField(
  CoordinatorInnerModule.DWalletCoordinatorInner
);
const SystemInnerDynamicField = DynamicField(SystemInnerModule.SystemInner);
export {
  CoordinatorInnerDynamicField,
  Curve,
  DWalletKind,
  DynamicField,
  Hash,
  SignatureAlgorithm,
  SystemInnerDynamicField
};
