// @ika.xyz/sdk/dist/cjs/client/errors.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var errors_exports = {};
__export(errors_exports, {
  CacheError: () => CacheError,
  IkaClientError: () => IkaClientError,
  InvalidObjectError: () => InvalidObjectError,
  NetworkError: () => NetworkError,
  ObjectNotFoundError: () => ObjectNotFoundError
});
module.exports = __toCommonJS(errors_exports);
class IkaClientError extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "IkaClientError";
  }
}
class ObjectNotFoundError extends IkaClientError {
  constructor(objectType, objectId, cause) {
    const idInfo = objectId ? ` with ID ${objectId}` : "";
    super(`${objectType} object${idInfo} not found`, cause);
    this.name = "ObjectNotFoundError";
  }
}
class InvalidObjectError extends IkaClientError {
  constructor(objectType, objectId, cause) {
    const idInfo = objectId ? ` (ID: ${objectId})` : "";
    super(`Invalid ${objectType} object${idInfo}: Expected structure not found`, cause);
    this.name = "InvalidObjectError";
  }
}
class CacheError extends IkaClientError {
  constructor(message, cause) {
    super(`Cache error: ${message}`, cause);
    this.name = "CacheError";
  }
}
class NetworkError extends IkaClientError {
  constructor(message, cause) {
    super(`Network error: ${message}`, cause);
    this.name = "NetworkError";
  }
}
