// @ika.xyz/sdk/dist/esm/generated/ika_dwallet_2pc_mpc/coordinator.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
const $moduleName = "@local-pkg/2pc-mpc::coordinator";
const DWalletCoordinator = new MoveStruct({ name: `${$moduleName}::DWalletCoordinator`, fields: {
  id: object.UID,
  version: bcs.u64(),
  package_id: bcs.Address,
  new_package_id: bcs.option(bcs.Address),
  migration_epoch: bcs.option(bcs.u64())
} });
function processCheckpointMessageByQuorum(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "vector<u8>",
    "vector<u8>",
    "vector<u8>"
  ];
  const parameterNames = ["dwallet_2pcMpcCoordinator", "signature", "signersBitmap", "message"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "process_checkpoint_message_by_quorum",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function initiateMidEpochReconfiguration(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self", "systemCurrentStatusInfo"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "initiate_mid_epoch_reconfiguration",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestNetworkEncryptionKeyMidEpochReconfiguration(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "dwalletNetworkEncryptionKeyId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_network_encryption_key_mid_epoch_reconfiguration",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function advanceEpoch(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::advance_epoch_approver::AdvanceEpochApprover`
  ];
  const parameterNames = ["self", "advanceEpochApprover"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "advance_epoch",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestDwalletNetworkEncryptionKeyDkgByCap(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "vector<u8>",
    `${packageAddress}::protocol_cap::VerifiedProtocolCap`
  ];
  const parameterNames = ["self", "paramsForNetwork", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_dwallet_network_encryption_key_dkg_by_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function processCheckpointMessageByCap(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "vector<u8>",
    `${packageAddress}::protocol_cap::VerifiedProtocolCap`
  ];
  const parameterNames = ["self", "message", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "process_checkpoint_message_by_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function setGasFeeReimbursementSuiSystemCallValueByCap(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "u64",
    `${packageAddress}::protocol_cap::VerifiedProtocolCap`
  ];
  const parameterNames = ["self", "gasFeeReimbursementSuiSystemCallValue", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "set_gas_fee_reimbursement_sui_system_call_value_by_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function setSupportedAndPricing(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::pricing::PricingInfo`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::vec_map::VecMap<u32, 0x0000000000000000000000000000000000000000000000000000000000000002::vec_map::VecMap<u32, vector<u32>>>",
    `${packageAddress}::protocol_cap::VerifiedProtocolCap`
  ];
  const parameterNames = ["self", "defaultPricing", "supportedCurvesToSignatureAlgorithmsToHashSchemes", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "set_supported_and_pricing",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function setPausedCurvesAndSignatureAlgorithms(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "vector<u32>",
    "vector<u32>",
    "vector<u32>",
    `${packageAddress}::protocol_cap::VerifiedProtocolCap`
  ];
  const parameterNames = ["self", "pausedCurves", "pausedSignatureAlgorithms", "pausedHashSchemes", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "set_paused_curves_and_signature_algorithms",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function setGlobalPresignConfig(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::vec_map::VecMap<u32, vector<u32>>",
    "0x0000000000000000000000000000000000000000000000000000000000000002::vec_map::VecMap<u32, vector<u32>>",
    `${packageAddress}::protocol_cap::VerifiedProtocolCap`
  ];
  const parameterNames = ["self", "curveToSignatureAlgorithmsForDkg", "curveToSignatureAlgorithmsForImportedKey", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "set_global_presign_config",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestLockEpochSessions(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::system_current_status_info::SystemCurrentStatusInfo`
  ];
  const parameterNames = ["self", "systemCurrentStatusInfo"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_lock_epoch_sessions",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function setPricingVote(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::pricing::PricingInfo`,
    `${packageAddress}::validator_cap::VerifiedValidatorOperationCap`
  ];
  const parameterNames = ["self", "pricing", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "set_pricing_vote",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function registerSessionIdentifier(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "vector<u8>"
  ];
  const parameterNames = ["self", "bytes"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "register_session_identifier",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function getActiveEncryptionKey(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "address"
  ];
  const parameterNames = ["self", "address"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "get_active_encryption_key",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function registerEncryptionKey(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "u32",
    "vector<u8>",
    "vector<u8>",
    "vector<u8>"
  ];
  const parameterNames = ["self", "curve", "encryptionKey", "encryptionKeySignature", "signerPublicKey"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "register_encryption_key",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function approveMessage(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::DWalletCap`,
    "u32",
    "u32",
    "vector<u8>"
  ];
  const parameterNames = ["self", "dwalletCap", "signatureAlgorithm", "hashScheme", "message"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "approve_message",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function approveImportedKeyMessage(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::ImportedKeyDWalletCap`,
    "u32",
    "u32",
    "vector<u8>"
  ];
  const parameterNames = ["self", "importedKeyDwalletCap", "signatureAlgorithm", "hashScheme", "message"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "approve_imported_key_message",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestDwalletDkgFirstRound(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u32",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["Self", "DwalletNetworkEncryptionKeyId", "Curve", "SessionIdentifier", "PaymentIka", "PaymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_dwallet_dkg_first_round",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestDwalletDkgSecondRound(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::DWalletCap`,
    "vector<u8>",
    "vector<u8>",
    "address",
    "vector<u8>",
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["Self", "DwalletCap", "CentralizedPublicKeyShareAndProof", "EncryptedCentralizedSecretShareAndProof", "EncryptionKeyAddress", "UserPublicOutput", "SignerPublicKey", "SessionIdentifier", "PaymentIka", "PaymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_dwallet_dkg_second_round",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function signDuringDkgRequest(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPresignCap`,
    "u32",
    "vector<u8>",
    "vector<u8>"
  ];
  const parameterNames = ["self", "presignCap", "hashScheme", "message", "messageCentralizedSignature"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "sign_during_dkg_request",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestDwalletDkg(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u32",
    "vector<u8>",
    "vector<u8>",
    "address",
    "vector<u8>",
    "vector<u8>",
    `0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<${packageAddress}::coordinator_inner::SignDuringDKGRequest>`,
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletNetworkEncryptionKeyId", "curve", "centralizedPublicKeyShareAndProof", "encryptedCentralizedSecretShareAndProof", "encryptionKeyAddress", "userPublicOutput", "signerPublicKey", "signDuringDkgRequest", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_dwallet_dkg",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestDwalletDkgWithPublicUserSecretKeyShare(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u32",
    "vector<u8>",
    "vector<u8>",
    "vector<u8>",
    `0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<${packageAddress}::coordinator_inner::SignDuringDKGRequest>`,
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletNetworkEncryptionKeyId", "curve", "centralizedPublicKeyShareAndProof", "userPublicOutput", "publicUserSecretKeyShare", "signDuringDkgRequest", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_dwallet_dkg_with_public_user_secret_key_share",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function calculatePricingVotes(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "u32",
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u32>",
    "u32"
  ];
  const parameterNames = ["self", "curve", "signatureAlgorithm", "protocol"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "calculate_pricing_votes",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestImportedKeyDwalletVerification(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u32",
    "vector<u8>",
    "vector<u8>",
    "address",
    "vector<u8>",
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletNetworkEncryptionKeyId", "curve", "centralizedPartyMessage", "encryptedCentralizedSecretShareAndProof", "encryptionKeyAddress", "userPublicOutput", "signerPublicKey", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_imported_key_dwallet_verification",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestMakeDwalletUserSecretKeySharesPublic(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletId", "publicUserSecretKeyShares", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_make_dwallet_user_secret_key_shares_public",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestReEncryptUserShareFor(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "address",
    "vector<u8>",
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletId", "destinationEncryptionKeyAddress", "encryptedCentralizedSecretShareAndProof", "sourceEncryptedUserSecretKeyShareId", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_re_encrypt_user_share_for",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function acceptEncryptedUserShare(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "vector<u8>"
  ];
  const parameterNames = ["self", "dwalletId", "encryptedUserSecretKeyShareId", "userOutputSignature"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "accept_encrypted_user_share",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestPresign(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u32",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletId", "signatureAlgorithm", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_presign",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestGlobalPresign(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    "u32",
    "u32",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletNetworkEncryptionKeyId", "curve", "signatureAlgorithm", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_global_presign",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isPresignValid(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::UnverifiedPresignCap`
  ];
  const parameterNames = ["self", "presignCap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "is_presign_valid",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifyPresignCap(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::UnverifiedPresignCap`
  ];
  const parameterNames = ["self", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "verify_presign_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestSign(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPresignCap`,
    `${packageAddress}::coordinator_inner::MessageApproval`,
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "presignCap", "messageApproval", "messageCentralizedSignature", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_sign",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestSignAndReturnId(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPresignCap`,
    `${packageAddress}::coordinator_inner::MessageApproval`,
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "presignCap", "messageApproval", "messageCentralizedSignature", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_sign_and_return_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestImportedKeySign(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPresignCap`,
    `${packageAddress}::coordinator_inner::ImportedKeyMessageApproval`,
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "presignCap", "messageApproval", "messageCentralizedSignature", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_imported_key_sign",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestImportedKeySignAndReturnId(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPresignCap`,
    `${packageAddress}::coordinator_inner::ImportedKeyMessageApproval`,
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "presignCap", "messageApproval", "messageCentralizedSignature", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_imported_key_sign_and_return_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestFutureSign(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
    `${packageAddress}::coordinator_inner::VerifiedPresignCap`,
    "vector<u8>",
    "u32",
    "vector<u8>",
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "dwalletId", "presignCap", "message", "hashScheme", "messageCentralizedSignature", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_future_sign",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function isPartialUserSignatureValid(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::UnverifiedPartialUserSignatureCap`
  ];
  const parameterNames = ["self", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "is_partial_user_signature_valid",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function verifyPartialUserSignatureCap(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::UnverifiedPartialUserSignatureCap`
  ];
  const parameterNames = ["self", "cap"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "verify_partial_user_signature_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestSignWithPartialUserSignature(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPartialUserSignatureCap`,
    `${packageAddress}::coordinator_inner::MessageApproval`,
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "partialUserSignatureCap", "messageApproval", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_sign_with_partial_user_signature",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestSignWithPartialUserSignatureAndReturnId(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPartialUserSignatureCap`,
    `${packageAddress}::coordinator_inner::MessageApproval`,
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "partialUserSignatureCap", "messageApproval", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_sign_with_partial_user_signature_and_return_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestImportedKeySignWithPartialUserSignature(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPartialUserSignatureCap`,
    `${packageAddress}::coordinator_inner::ImportedKeyMessageApproval`,
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "partialUserSignatureCap", "messageApproval", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_imported_key_sign_with_partial_user_signature",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function requestImportedKeySignWithPartialUserSignatureAndReturnId(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPartialUserSignatureCap`,
    `${packageAddress}::coordinator_inner::ImportedKeyMessageApproval`,
    `${packageAddress}::sessions_manager::SessionIdentifier`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "partialUserSignatureCap", "messageApproval", "sessionIdentifier", "paymentIka", "paymentSui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "request_imported_key_sign_with_partial_user_signature_and_return_id",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function matchPartialUserSignatureWithMessageApproval(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPartialUserSignatureCap`,
    `${packageAddress}::coordinator_inner::MessageApproval`
  ];
  const parameterNames = ["self", "partialUserSignatureCap", "messageApproval"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "match_partial_user_signature_with_message_approval",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function matchPartialUserSignatureWithImportedKeyMessageApproval(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::coordinator_inner::VerifiedPartialUserSignatureCap`,
    `${packageAddress}::coordinator_inner::ImportedKeyMessageApproval`
  ];
  const parameterNames = ["self", "partialUserSignatureCap", "messageApproval"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "match_partial_user_signature_with_imported_key_message_approval",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function hasDwallet(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "dwalletId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "has_dwallet",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function getDwallet(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
  ];
  const parameterNames = ["self", "dwalletId"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "get_dwallet",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function currentPricing(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "current_pricing",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function subsidizeCoordinatorWithSui(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    "0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI>"
  ];
  const parameterNames = ["self", "sui"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "subsidize_coordinator_with_sui",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function subsidizeCoordinatorWithIka(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::ika::IKA>`
  ];
  const parameterNames = ["self", "ika"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "subsidize_coordinator_with_ika",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function commitUpgrade(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::upgrade_package_approver::UpgradePackageApprover`
  ];
  const parameterNames = ["self", "upgradePackageApprover"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "commit_upgrade",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function tryMigrateByCap(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`,
    `${packageAddress}::protocol_cap::VerifiedProtocolCap`
  ];
  const parameterNames = ["self", "_"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "try_migrate_by_cap",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function tryMigrate(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "try_migrate",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function version(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::coordinator::DWalletCoordinator`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "coordinator",
    function: "version",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  DWalletCoordinator,
  acceptEncryptedUserShare,
  advanceEpoch,
  approveImportedKeyMessage,
  approveMessage,
  calculatePricingVotes,
  commitUpgrade,
  currentPricing,
  getActiveEncryptionKey,
  getDwallet,
  hasDwallet,
  initiateMidEpochReconfiguration,
  isPartialUserSignatureValid,
  isPresignValid,
  matchPartialUserSignatureWithImportedKeyMessageApproval,
  matchPartialUserSignatureWithMessageApproval,
  processCheckpointMessageByCap,
  processCheckpointMessageByQuorum,
  registerEncryptionKey,
  registerSessionIdentifier,
  requestDwalletDkg,
  requestDwalletDkgFirstRound,
  requestDwalletDkgSecondRound,
  requestDwalletDkgWithPublicUserSecretKeyShare,
  requestDwalletNetworkEncryptionKeyDkgByCap,
  requestFutureSign,
  requestGlobalPresign,
  requestImportedKeyDwalletVerification,
  requestImportedKeySign,
  requestImportedKeySignAndReturnId,
  requestImportedKeySignWithPartialUserSignature,
  requestImportedKeySignWithPartialUserSignatureAndReturnId,
  requestLockEpochSessions,
  requestMakeDwalletUserSecretKeySharesPublic,
  requestNetworkEncryptionKeyMidEpochReconfiguration,
  requestPresign,
  requestReEncryptUserShareFor,
  requestSign,
  requestSignAndReturnId,
  requestSignWithPartialUserSignature,
  requestSignWithPartialUserSignatureAndReturnId,
  setGasFeeReimbursementSuiSystemCallValueByCap,
  setGlobalPresignConfig,
  setPausedCurvesAndSignatureAlgorithms,
  setPricingVote,
  setSupportedAndPricing,
  signDuringDkgRequest,
  subsidizeCoordinatorWithIka,
  subsidizeCoordinatorWithSui,
  tryMigrate,
  tryMigrateByCap,
  verifyPartialUserSignatureCap,
  verifyPresignCap,
  version
};
