// @ika.xyz/sdk/dist/esm/generated/ika_dwallet_2pc_mpc/pricing.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as vec_map from "./deps/sui/vec_map.js";
import * as bls_committee from "./deps/ika_common/bls_committee.js";
const $moduleName = "@local-pkg/2pc-mpc::pricing";
const PricingInfoKey = new MoveStruct({ name: `${$moduleName}::PricingInfoKey`, fields: {
  curve: bcs.u32(),
  signature_algorithm: bcs.option(bcs.u32()),
  protocol: bcs.u32()
} });
const PricingInfoValue = new MoveStruct({ name: `${$moduleName}::PricingInfoValue`, fields: {
  fee_ika: bcs.u64(),
  gas_fee_reimbursement_sui: bcs.u64(),
  gas_fee_reimbursement_sui_for_system_calls: bcs.u64()
} });
const PricingInfo = new MoveStruct({ name: `${$moduleName}::PricingInfo`, fields: {
  /**
     * The pricing for each curve and signature algorithm and protocol. The first key
     * is the curve, the second is the signature algorithm, the third is the protocol.
     */
  pricing_map: vec_map.VecMap(PricingInfoKey, PricingInfoValue)
} });
const PricingInfoCalculationVotes = new MoveStruct({ name: `${$moduleName}::PricingInfoCalculationVotes`, fields: {
  bls_committee: bls_committee.BlsCommittee,
  default_pricing: PricingInfo,
  working_pricing: PricingInfo
} });
function empty(options = {}) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "empty"
  });
}
function insertOrUpdatePricing(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfo`,
    "u32",
    "0x0000000000000000000000000000000000000000000000000000000000000001::option::Option<u32>",
    "u32",
    "u64",
    "u64",
    "u64"
  ];
  const parameterNames = ["self", "curve", "signatureAlgorithm", "protocol", "feeIka", "gasFeeReimbursementSui", "gasFeeReimbursementSuiForSystemCalls"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "insert_or_update_pricing",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function feeIka(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfoValue`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "fee_ika",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function gasFeeReimbursementSui(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfoValue`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "gas_fee_reimbursement_sui",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function gasFeeReimbursementSuiForSystemCalls(options) {
  const packageAddress = options.package ?? "@local-pkg/2pc-mpc";
  const argumentsTypes = [
    `${packageAddress}::pricing::PricingInfoValue`
  ];
  const parameterNames = ["self"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "pricing",
    function: "gas_fee_reimbursement_sui_for_system_calls",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  PricingInfo,
  PricingInfoCalculationVotes,
  PricingInfoKey,
  PricingInfoValue,
  empty,
  feeIka,
  gasFeeReimbursementSui,
  gasFeeReimbursementSuiForSystemCalls,
  insertOrUpdatePricing
};
